src/plots/barPlot.ts

```diff
@@ -468,78 +468,119 @@ export module Plots {
-        let x = attrToProjector["x"](d, i, dataset);
-        let y = attrToProjector["y"](d, i, dataset);
-        let positive = originalPositionFn(d, i, dataset) <= scaledBaseline;
+
+      let drawLabel = (d: any, i: number) => {
+        let valueAccessor = this._isVertical ? this.y().accessor : this.x().accessor;
+        let value = valueAccessor(d, i, dataset);
+        let valueScale: Scale<any, number> = this._isVertical ? this.y().scale : this.x().scale;
+        let scaledValue = valueScale != null ? valueScale.scale(value) : value;
+        let scaledBaseline = valueScale != null ? valueScale.scale(this.baselineValue()) : this.baselineValue();
+
+        let barWidth = attrToProjector["width"](d, i, dataset);
+        let barHeight = attrToProjector["height"](d, i, dataset);
+        let text = this._labelFormatter(valueAccessor(d, i, dataset)).toString();
         let measurement = measurer.measure(text);
-        let color = attrToProjector["fill"](d, i, dataset);
-        let dark = Utils.Color.contrast("white", color) * 1.6 < Utils.Color.contrast("black", color);
-        let primary = this._isVertical ? h : w;
-        let primarySpace = this._isVertical ? measurement.height : measurement.width;
-        let secondaryAttrTextSpace = this._isVertical ? measurement.width : measurement.height;
-        let secondaryAttrAvailableSpace = this._isVertical ? w : h;
-        let tooWide = secondaryAttrTextSpace + 2 * Bar._LABEL_HORIZONTAL_PADDING > secondaryAttrAvailableSpace;
-        if (measurement.height <= h && measurement.width <= w) {
-          let offset = Math.min((primary - primarySpace) / 2, Bar._LABEL_VERTICAL_PADDING);
-          if (!positive) { offset = offset * -1; }
-          if (this._isVertical) {
-            y += offset;
-          } else {
-            x += offset;
+
+        let xAlignment = "center";
+        let yAlignment = "center";
+        let labelContainerOrigin = {
+          x: attrToProjector["x"](d, i, dataset),
+          y: attrToProjector["y"](d, i, dataset)
+        };
+        let containerWidth = barWidth;
+        let containerHeight = barHeight;
+
+        let labelOrigin = {
+          x: labelContainerOrigin.x,
+          y: labelContainerOrigin.y
+        };
+
+        let showLabelOnBar = this._isVertical ? (measurement.height <= barHeight) : (measurement.width <= barWidth);
+
+        if (this._isVertical) {
+          labelOrigin.x += containerWidth / 2 - measurement.width / 2;
+
+          if (showLabelOnBar) {
+            let offset = Math.min( (barHeight - measurement.height ) / 2, Bar._LABEL_VERTICAL_PADDING);
+            if (scaledValue < scaledBaseline) {
+              labelContainerOrigin.y += offset;
+              yAlignment = "top";
+              labelOrigin.y += offset;
+            } else {
+              labelContainerOrigin.y -= offset;
+              yAlignment = "bottom";
+              labelOrigin.y += containerHeight - offset - measurement.height;
+            }
```

Comment:
The `if`-block implementation here is intended to make the distinct cases for labels clearer:
* `Plot` is vertical or horizontal?
* label is on-bar or off-bar?
* value is positive or negative?

The original code used ternaries, but it was sometimes difficult to understand which case a particular line of logic was being written for.